Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs

```diff
@@ -17,234 +17,72 @@
-      // 
-      // (null, null) -> ignore
-      // (null, *)    -> scenario a)
-      // (*, null)    -> remember * and set special context
-      // (*, *)       -> scenario a) if no special context, otherwise b)
-
-      StorageContext newStorageContext;
-      if (previousContainer == null)
-      {
-        // (null, null) -> ignore
-        if (currentContainer == null)
-          return;
-
-        // (null, *)    -> scenario a)
-        newStorageContext = StorageContext.Empty;
-      }
-      else
-      {
-        // (*, null)    -> remember * and set special context
-        if (currentContainer == null)
-        {
-          newStorageContext = StorageContext.FromPrevious(previousContainer);
-        }
-        // (*, *)       -> scenario a) if no special context, otherwise b)
-        else
-        {
-          newStorageContext = previousContainer.Previous == currentContainer
-              ? currentContainer
-              : StorageContext.FromPrevious(previousContainer);
-        }
-      }
-
-      // Update only if the container changed to prevent unnecessary updates
-      if (currentContainer != newStorageContext)
-        _context.Value = newStorageContext;
+      return new SafeContextBoundary(this, hashtable);
     }
 
     /// <inheritdoc />
     public object? GetData (string key)
     {
-      ArgumentUtility.CheckNotNull("key", key);
+      ArgumentUtility.CheckNotNull(nameof(key), key);
 
-      return _context.Value?.GetData(key);
+      return _context.Value?[key];
     }
 
     /// <inheritdoc />
     public void SetData (string key, object? value)
     {
-      ArgumentUtility.CheckNotNull("key", key);
+      ArgumentUtility.CheckNotNull(nameof(key), key);
 
 #if DEBUG && NETFRAMEWORK
       if (value is System.Runtime.Remoting.Messaging.ILogicalThreadAffinative)
         throw new NotSupportedException("Remoting is not supported.");
 #endif
 
-      var oldValue = _context.Value;
-      var newValue = (oldValue ?? StorageContext.Empty).SetData(key, value);
-      if (oldValue != newValue)
-        _context.Value = newValue;
+      var hashtable = _context.Value;
+      if (hashtable == null)
+      {
+        hashtable = Hashtable.Synchronized(new Hashtable());
+        _context.Value = hashtable;
+      }
+
+      hashtable[key] = value;
     }
 
     /// <inheritdoc />
     public void FreeData (string key)
     {
-      ArgumentUtility.CheckNotNull("key", key);
+      ArgumentUtility.CheckNotNull(nameof(key), key);
 
-      var oldValue = _context.Value;
-      if (oldValue == null)
-        return;
+      _context.Value?.Remove(key);
+    }
 
-      var newValue = oldValue.FreeData(key);
-      if (oldValue != newValue)
-        _context.Value = newValue;
+    /// <inheritdoc />
+    void ISafeContextBoundaryStrategy.RestorePreviousSafeContext (object? state)
+    {
+      _context.Value = state as Hashtable;
```

Comment:
Hard cast.

Soft cast is only used when a type test is explicitly intended and a mismatched type does not indicate a bug. Here, an invalid state type will mean something's gone horribly wrong.